### PR TITLE
osd: fix dump_ops_in_flight races

### PR DESCRIPTION
**Pull request by dzafman**
Fixes: #8885

### CODE REVIEW

**Comment by dzafman** [comment]
Haven't written an automated test case, yet.  Rebased to jewel so waiting for my own make check to verify.

File: src/messages/MOSDOp.h [hunk 1]
```diff
@@ -48,8 +48,10 @@ class MOSDOp : public Message {
   pg_t pgid;
   bufferlist::iterator p;
   // Decoding flags. Decoding is only needed for messages catched by pipe reader.
-  bool partial_decode_needed;
-  bool final_decode_needed;
+  // Transition from true -> false without locks being held
+  // Can never see final_decode_needed == false and partial_decode_needed == true
+  volatile bool partial_decode_needed;
+  volatile bool final_decode_needed;
```
Comment (gregsfortytwo) [inline src/messages/MOSDOp.h]: I'm just looking up the volatile keyword here, but I'm not sure this works, strictly speaking. http://en.cppreference.com/w/cpp/language/cv and https://en.wikipedia.org/wiki/Volatile_(computer_programming)#In_C_and_C.2B.2B are both pretty clear that ```volatile``` is not "for inter-thread communication"

Comment (tchaikov) [inline src/messages/MOSDOp.h]: agreed, can we use `atomic<bool>` instead here?

Comment (gregsfortytwo) [inline src/messages/MOSDOp.h]: I think that'll get expensive; don't we dereference those bools every time we look at an MOSDOp? I *think* we're just using those as flags to determine if we need to grab a lock and do the decode before looking at memory, or if we can look at the memory directly, which means presumably with careful coding it's okay if we read a stale value (we'd just grab the lock needlessly). But apparently it's being done wrongly, or maybe it really is just that the memory reads are being reordered incorrectly (which is what ```volatile``` prevents)...

Comment (athanatos) [inline src/messages/MOSDOp.h]: Yeah, volatile really isn't appropriate here.  atomic<bool> would be cleanest.

File: src/common/TrackedOp.h [hunk 2]
```diff
@@ -166,13 +170,15 @@ class TrackedOp {
...
+    is_registered(false)
   {
     RWLock::RLocker l(tracker->lock);
     if (tracker->tracking_enabled) {
       tracker->register_inflight_op(&xitem);
       events.push_back(make_pair(initiated_at, "initiated"));
-      is_tracked = true;
+      // Tells tracking_start() whether to set is_tracked or not
+      is_registered = true;
     }
```
Comment (athanatos) [inline src/common/TrackedOp.h]: Hmm, this does solve the problem, but I think the better solution would be to move this code into a TrackedOp::register method and call it in OpTracker::create_request.  That way, we never have a half-constructed object registered in the first place and can avoid having both is_tracked and is_registered.

Comment (athanatos) [inline src/common/TrackedOp.h]: That is, we can have just is_tracked rather than both.

Comment (dzafman) [inline src/common/TrackedOp.h]: I have the fix for this by moving that constructor code into tracking_start() so that is_registered isn't needed.

**Comment by athanatos** [comment]
lgtm

**Comment by liewegas** [comment]
passed my rados run

**Pull request merged**